docs/advanced-performance-analysis.md

```diff
@@ -0,0 +1,226 @@
+                36      context-switches          #    0.080 K/sec                  
+                 9      cpu-migrations            #    0.020 K/sec                  
+              2989      page-faults               #    0.007 M/sec                  
+        1736698259      cycles                    #    3.860 GHz                    
+          53504877      stalled-cycles-frontend   #    3.08% frontend cycles idle   
+        5927885177      instructions              #    3.41  insn per cycle         
+                                                  #    0.01  stalled cycles per insn
+        1422566509      branches                  # 3161.555 M/sec                  
+            375326      branch-misses             #    0.03% of all branches        
+
+       0.449450377 seconds time elapsed
+
+       0.442788000 seconds user
+       0.008050000 seconds sys
+```
+
+First, we can confirm that it is nicely CPU bound (`1.001 CPUs utilized`).
+
+We can also see that we executed about 5.9B instructions during this benchmark run. The number of instructions executed itself is not very meaningful but it can be very useful when comparing between versions of the same code. For example, let's assume we changed some code and want to confirm in a benchmark that indeed we have made it faster, then it can be very useful to see that we have also reduced the number of instructions.
+
+With `perf` you could run `perf stat -e instructions -- ./benchmark-before-change` and `perf stat -e instructions -- ./benchmark-after-change` and confirm that:
+
+1. The runtime (_time elapsed_) has gone down and
+2. the number of instructions executed has gone down too.
+
+Note that the time it takes to run an instruction varies wildly so it is in theory possible to reduce the number of instructions but make performance actually worse. An example would be if you traded a lot of simple arithmetic for memory accesses. Said that, in most cases where you reduced the number of instructions significantly, you will also see an improvement in the actual runtime.
+
+Another place where instruction counts come in very handy is when you see this maybe impossible-sounding situation: the run time varies significantly but the number of executed instructions is _the same_. The motivating example is exactly this situation: The code executed is the _same_, confirmed by manually checking the generated assembly for the code run as well as checking that the number of instructions is almost equal. Because manual assembly analysis can easily go wrong, it's advisable to compare the number of executed instructions even if you convinced yourself that the executed code is the same. Of course, there are other options like single-stepping in a debugger but that gets tedious quickly. If the executed code and the inputs are the same, then the number of executed instructions should also be the same (ie. < 0.01% variance).
+
+## CPU stalls
```

Comment:
How did you make the jump from runtime difference and the same instruction count to CPU stalls? Was there anything else you looked at at this point before looking at stalls?

Reply:
@glbrntt Thanks! I'll flesh that out a little. But in general: If I know that the instructions run are the same but one is faster, then it needs to be some kind of stall/penalty really. But I'll add some text.